lgc/builder/InOutBuilder.cpp

```diff
@@ -65,6 +65,162 @@ Value *InOutBuilder::CreateReadGenericInput(Type *resultTy, unsigned location, V
                                 vertexIndex, instName);
 }
 
+// =====================================================================================================================
+// Create a read of (part of) a perVertex input value, passed from the previous shader stage.
+// The result type is as specified by pResultTy, a scalar or vector type with no more than four elements.
+// A "location" contains four "components", each of which can contain a 16- or 32-bit scalar value. A
+// 64-bit scalar value takes two components.
+//
+// @param resultTy : Type of value to read
+// @param location : Base location (row) of input
+// @param locationOffset : Variable location offset; must be within locationCount
+// @param elemIdx : Element index in vector. (This is the SPIR-V "component", except that it is half the component for
+// 64-bit elements.)
+// @param locationCount : Count of locations taken by the input
+// @param inputInfo : Extra input info (FS interp info)
+// @param vertexIndex : Vertex index; for FS custom interpolated input: auxiliary interpolation value;
+// @param instName : Name to give instruction(s)
+// @returns Value of input
+Value *InOutBuilder::CreateReadPerVertexInput(Type *resultTy, unsigned location, Value *locationOffset, Value *elemIdx,
+                                              unsigned locationCount, InOutInfo inputInfo, Value *vertexIndex,
+                                              const Twine &instName) {
+  assert(!resultTy->isAggregateType());
+  assert(inputInfo.getInterpMode() == InOutInfo::InterpModeCustom);
+  assert(m_shaderStage == ShaderStageFragment);
+
+  // Fold constant pLocationOffset into location.
+  if (auto constLocOffset = dyn_cast<ConstantInt>(locationOffset)) {
+    location += constLocOffset->getZExtValue();
+    locationOffset = getInt32(0);
+    locationCount = (resultTy->getPrimitiveSizeInBits() + 127U) / 128U;
+  }
+
+  // Mark the usage of the input/output.
+  markGenericInputOutputUsage(false, location, locationCount, inputInfo, vertexIndex);
+
+  // Fixing the order
+  bool parity = false;
+  Value *odd = nullptr;
+  Value *even = nullptr;
+
+  auto primType = getPipelineState()->getPrimitiveType();
+  switch (primType) {
+  case PrimitiveType::Point:
+    break;
+  case PrimitiveType::Line_List:
+  case PrimitiveType::Line_Strip:
+    break;
+  case PrimitiveType::Triangle_List: {
+    switch (dyn_cast<ConstantInt>(vertexIndex)->getZExtValue()) {
+    case 0: {
+      vertexIndex = getInt32(2);
+      break;
+    }
+    case 1: {
+      vertexIndex = getInt32(0);
+      break;
+    }
+    case 2: {
+      vertexIndex = getInt32(1);
+      break;
+    }
+    }
+    break;
+  }
+  case PrimitiveType::Triangle_Strip:
+  case PrimitiveType::Triangle_Fan:
+  case PrimitiveType::Triangle_Strip_Adjacency: {
+    switch (dyn_cast<ConstantInt>(vertexIndex)->getZExtValue()) {
```

Comment:
This can just be a `cast`. If the `dyn_cast` returned `null`, it would crash anyway.
(The same goes for the cast above and below.)